docs/index.rst

```diff
@@ -1,260 +1,427 @@
+(`http_proxy`, `https_proxy`, `no_proxy`, etc). Libraries can use all the
+facilities of Confit (with the exception of ``from_args``) to ease handling
+of their configuration.
+
+
+Configuration
+-------------
+
+A configuration consists of variable bindings. In each new project, a
+configuration is born when some variable holds an arbitrary value, e.g.  a
+path to a temporary directory. The value is moved to the configuration as a
+default, and the variable's value is loaded from the configuration.
+
+Variables
+^^^^^^^^^
+
+In Confit, a configuration is an immutable mapping from string names to
+*variables*. Variables have two methods, ``get`` and ``origin``.
+
+::
+
+  get(cast=lambda x: x)
+
+Call *cast* with the variable's value and return the result. *cast* is useful
+for casting the value to a different type (i.e., "get the value as a ___").
+*cast* may also check the value against some validity criteria, e.g. whether
+a string can be parsed, or whether an integer fits within a range. If *cast*
+raises an exception, then it will be wrapped in a ``ConfitCastError``, that,
+if uncaught, will be pretty-printed to the console (using ``sys.excepthook``)
+indicating both the failing call to ``get`` and the origin of the faulty
+value::
+
+  /usr/bin/script:123:
+      port = config['port'].get(is_not_privileged)
+  failed on the override at /home/user/.example.yml:8:
+  port: 80
+
+::
+
+  origin()
+
+Return the variable's origin. An **origin** is a location identifier for an
+individual variable within a given source (see Sources section). Each origin
+type may have different properties, but every built-in origin can be
+pretty-printed with its ``__str__`` method.
+
+
+Sources
+-------
+
+Configurations come from **sources**, e.g. command-line arguments,
+environment variables, or files. Confit comes with a few functions for
+creating configurations from well-known sources.
+
+::
+
+  from_mapping(mapping)
+
+Return a configuration taken from a mapping from string names to values. Each
+variable will have a ``FileOrigin`` pointing to the call to this function.
+
+::
+
+  from_object(object)
+
+Return a configuration drawn from the non-callable public properties of
+*object*. A property is considered public if its name does not start with an
+underscore. Each variable will have a ``FileOrigin`` pointing to the call to
+this function.
+
+::
+
+  from_yaml(path)
+
+Return a configuration parsed from a YAML file at *path*, or if *path* is not
+a file, return an empty configuration. Each variable will have a
+``FileOrigin``. Can be used to parse JSON as well since YAML is a superset.
+
+::
+
+  from_ini(path, sep="_")
+
+Return a configuration parsed from an INI file at *path*, or if *path* is not
+a file, return an empty configuration. Each variable will have a
+``FileOrigin``, and its key will be prefixed by its section name separated by
+*sep*.
+
+::
+
+  user_config_paths(*paths)
```

Comment:
Organizational note: perhaps the `from_*` functions, which return configurations, should get a separate section from these other source-finding utilities.